### PR TITLE
Smoke test for F#

### PR DESCRIPTION
**Pull request by dsyme**
This addresses the first part of #540, i.e. adding an initial smoke test for ML.NET and F#

Submitting now to check that CI passes ok, will then iterate.

### CODE REVIEW

**Comment by dsyme** [comment]
Currently getting this on OSX, I don't yet understand why:
```
FSC : error FS2014: A problem occurred writing the binary '/Users/dotnet-bot/j/w/dotnet_machinelearning/master/osx10.13_debug_prtest/bin/obj/AnyCPU.Debug/Microsoft.ML.FSharp.Tests/netcoreapp2.0/Microsoft.ML.FSharp.Tests.dll': A call to StrongNameSignatureSize failed (Invalid Public Key blob) 
```

**Comment by dsyme** [comment]
This is ready for review etc.

Felt like a lot of work for +196 lines.  But still, there it is

File: test/Microsoft.ML.FSharp.Tests/SmokeTests.fs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        val mutable Sentiment : float32
+
+    type SentimentPrediction() =
+        [<ColumnName "PredictedLabel"; DefaultValue>]
+        val mutable Sentiment : bool
+
+    [<Fact>]
+    let ``FSharp-Sentiment-Smoke-Test`` () =
+
+        let testDataPath = __SOURCE_DIRECTORY__ + @"/../data/wikipedia-detox-250-line-data.tsv"
```
Comment (eerhardt) [inline test/Microsoft.ML.FSharp.Tests/SmokeTests.fs]: Does `__SOURCE_DIRECTORY__ ` work in non-interactive mode?

File: test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp2.0</TargetFrameworks>
+    <NoWarn>2003;$(NoWarn)</NoWarn>
+    <TargetFrameworks Condition="'$(OS)' != 'Unix'">$(TargetFrameworks); net461</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (dsyme) [reply test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: thanks!

File: test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    <PublicSign>false</PublicSign>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="SmokeTests.fs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.7.0" />
```
[the remark this answers is not in the document]
Comment (dsyme) [reply test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: ok, thanks!

File: test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    <PackageReference Include="xunit" Version="2.3.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
+    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
```
[the remark this answers is not in the document]
Comment (dsyme) [reply test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: okey dokey

File: test/Microsoft.ML.FSharp.Tests/SmokeTests.fs (hunk 1 above)
Comment (dsyme) [reply test/Microsoft.ML.FSharp.Tests/SmokeTests.fs]: yes it does

File: test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj [hunk 5]
```diff
@@ -0,0 +1,50 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Unix'">$(TargetFrameworks); net461</TargetFrameworks>
+    <NoWarn>2003;$(NoWarn)</NoWarn>
+    <PublicSign>false</PublicSign>
+    <SourceLink></SourceLink>
```
Comment (eerhardt) [inline test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: What's this for?  Is this necessary?

Comment (dsyme) [reply test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: I just got a build error when building from the IDE using VS 15.7 because for .NET 4.6.1 compilation "full" PDB symbols (not portable) were being produced), I can reproduce if I build with `msbuild` not `dotnet`.  Full PDB are not compatible with sourcelink. It doesn't go away even if you set DebugType to "portable".  

I think it is an F# build targets bug fixed in 15.8.   Anyway for now seemed sensible to turn of SourceLink for the test project

**Comment by dsyme** [comment]
OK, test is failing now it is running, oddly it's not failing on my machine.  Will look into it

```
2018-07-30T19:44:52.3992357Z Failed   Microsoft.ML.FSharp.Tests.SmokeTest1.FSharp-Sentiment-Smoke-Test
2018-07-30T19:44:52.3992813Z Error Message:
2018-07-30T19:44:52.3993183Z  System.InvalidOperationException : Entry point 'Transforms.TextFeaturizer' not found
2018-07-30T19:44:52.3993578Z Stack Trace:
2018-07-30T19:44:52.3994281Z    at Microsoft.ML.Runtime.EntryPoints.EntryPointNode..ctor(IHostEnvironment env, IChannel ch, ModuleCatalog moduleCatalog, RunContext context, String id, String entryPointName, JObject inputs, JObject outputs, Boolean checkpoint, String stageId, Single cost, String label, String group, String weight, String name)
2018-07-30T19:44:52.3994761Z    at Microsoft.ML.Runtime.EntryPoints.EntryPointNode.ValidateNodes(IHostEnvironment env, RunContext context, JArray nodes, ModuleCatalog moduleCatalog, String label, String group, String weight, String name)
2018-07-30T19:44:52.3995085Z    at Microsoft.ML.Runtime.EntryPoints.EntryPointGraph..ctor(IHostEnvironment env, ModuleCatalog moduleCatalog, JArray nodes)
2018-07-30T19:44:52.3995384Z    at Microsoft.ML.Runtime.Experiment.Compile()
2018-07-30T19:44:52.3995640Z    at Microsoft.ML.LearningPipeline.Train[TInput,TOutput]()
2018-07-30T19:44:52.3995938Z    at Microsoft.ML.FSharp.Tests.SmokeTest1.FSharp-Sentiment-Smoke-Test()
```

At least that's proof that it's running

**Comment by dsyme** [comment]
@eerhardt Is there any ongoing discussion about the design of ML.NET's Component Catalog mechanism, e.g. is there any prospect of scrapping it in favour of more traditional direct static library references?  What's the driving motivation of the mechanism?

I find it strange that we would have this extra level of indirection and it seems to be causing problems for F# - in scripts we need the artificial stuff mentioned in https://github.com/dotnet/machinelearning/issues/401.

Also, the component catalog seems to be populated via the assemblies reachable via direct static references (I'm not totally sure about that, but  that's kind of the point - the mechanism is just obscure).  But it's perfectly possible to write code there there isn't a static reference to, say `Microsoft.ML.Transforms`, e.g. prior to adding https://github.com/dotnet/machinelearning/pull/600/commits/4e131b5e4fadf7c9587707cbade211b70d0f6b8e there was no static reference to `Microsoft.ML.Transforms` in the F# unit test DLL.  I think (not yet sure) that was the cause of the unit test failure above.

Basically, what is the component catalog really for, and is it really necessary?  It's always going to sit very awkwardly alongside other .NET componentization models (including the traditional one of static references)  - e.g. using it with DI frameworks might be really rough.  Why isn't ML.NET just a normal set of .NET libraries using normal static assembly references?

**Comment by Ivanidzo4ka** [comment]
#208 is probably closest one. But it doesn't answer question why we have it in first place.

I can give you one example where existence of component catalog make sense (at least for me).

Simple scoring.
You got model trained somewhere and all you want is just run it.

With current code, you just do 
```
model.Read("model.zip");
model.Predict(data);
```
and all you need to do is add nuget package and it will work. (at least that was the case for .net framework 4.5 where content of nuget files was put into you binary folder, in net standard we need to make changes in our component catalog)

This way you don't have to add references on libraries in your project.


---
In reply to: [409231443](https://github.com/dotnet/machinelearning/pull/600#issuecomment-409231443) [](ancestors = 409231443)

**Comment by eerhardt** [comment]
Yes, there has been discussion here:  #371.  Basically, the plan of record is that we will move to directly calling the underlying C# APIs without going through the extra layer of indirection. That will solve the majority of the issues mentioned in #401.  (Note, we have the same issue in Azure Functions - #559)

However, that doesn't completely solve the issue of dependency injection. The other place where is must be used is when we load a `.zip` file (i.e. a "model" file) from disk. Someone could create a model file that used `EricsCoolLearner` from `EricsCool.dll`, and you can write an app that tries to load that model.  However, you don't necessarily have a direct/static library reference on `EricsCool.dll`.  (Or even if you did have the reference, you might not have used a type in that assembly yet, so it may not be loaded.)

So, in this scenario, we are still going to need dependency injection at runtime. That's the main usage scenario for component catalog.  (It is also used by other tooling, like the MAML command line tool, and the Entry Points subsystem.)

In investigating the Azure Functions issue, we discovered other issues with the way that the component catalog loads assemblies. At its heart, the issue is that we don't work well with custom AssemblyLoadContext instances.  That issue is being tracked at #569.

**Comment by dsyme** [comment]
Was integrated in https://github.com/dotnet/machinelearning/pull/616
